From e8fe8cb5f3f0a122fd59d67175d7c58e77dc0245 Mon Sep 17 00:00:00 2001
From: Martino Facchin <m.facchin@arduino.cc>
Date: Wed, 31 Jul 2019 12:48:04 +0200
Subject: [PATCH 01/31] HACK: avoid #10862 by not firing the assert

---
 drivers/source/usb/USBDevice.cpp | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/drivers/source/usb/USBDevice.cpp b/drivers/source/usb/USBDevice.cpp
index 6d3cfc87d4..ddeabcfa15 100644
--- a/drivers/source/usb/USBDevice.cpp
+++ b/drivers/source/usb/USBDevice.cpp
@@ -980,10 +980,11 @@ void USBDevice::out(usb_ep_t endpoint)
 
     endpoint_info_t *info = &_endpoint_info[EP_TO_INDEX(endpoint)];
 
-    MBED_ASSERT(info->pending >= 1);
-    info->pending -= 1;
-    if (info->callback) {
-        info->callback();
+    if (info->pending >= 1) {
+      info->pending -= 1;
+        if (info->callback) {
+          info->callback();
+        }
     }
 }
 
@@ -1007,10 +1008,11 @@ void USBDevice::in(usb_ep_t endpoint)
 
     endpoint_info_t *info = &_endpoint_info[EP_TO_INDEX(endpoint)];
 
-    MBED_ASSERT(info->pending >= 1);
-    info->pending -= 1;
-    if (info->callback) {
-        info->callback();
+    if (info->pending >= 1) {
+      info->pending -= 1;
+        if (info->callback) {
+          info->callback();
+        }
     }
 }
 
-- 
2.27.0

