From 02d37a845bf14acca743256d38b6005b54ff7e1b Mon Sep 17 00:00:00 2001
From: Martino Facchin <m.facchin@arduino.cc>
Date: Fri, 4 Oct 2019 15:52:47 +0200
Subject: [PATCH 03/31] QSPIFBlockDevice: use mbed namespace

---
 .../storage/blockdevice/COMPONENT_QSPIF/QSPIFBlockDevice.h    | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/components/storage/blockdevice/COMPONENT_QSPIF/QSPIFBlockDevice.h b/components/storage/blockdevice/COMPONENT_QSPIF/QSPIFBlockDevice.h
index 36c677b319..cacf0bc1b8 100644
--- a/components/storage/blockdevice/COMPONENT_QSPIF/QSPIFBlockDevice.h
+++ b/components/storage/blockdevice/COMPONENT_QSPIF/QSPIFBlockDevice.h
@@ -53,7 +53,7 @@
  */
 enum qspif_bd_error {
     QSPIF_BD_ERROR_OK                    = 0,     /*!< no error */
-    QSPIF_BD_ERROR_DEVICE_ERROR          = BD_ERROR_DEVICE_ERROR, /*!< device specific error -4001 */
+    QSPIF_BD_ERROR_DEVICE_ERROR          = mbed::BD_ERROR_DEVICE_ERROR, /*!< device specific error -4001 */
     QSPIF_BD_ERROR_PARSING_FAILED        = -4002, /* SFDP Parsing failed */
     QSPIF_BD_ERROR_READY_FAILED          = -4003, /* Wait for Mem Ready failed */
     QSPIF_BD_ERROR_WREN_FAILED           = -4004, /* Write Enable Failed */
@@ -289,7 +289,7 @@ private:
     qspi_status_t _qspi_set_frequency(int freq);
 
     // Update the 4-byte addressing extension register with the MSB of the address if it is in use
-    qspi_status_t _qspi_update_4byte_ext_addr_reg(bd_addr_t addr);
+    qspi_status_t _qspi_update_4byte_ext_addr_reg(mbed::bd_addr_t addr);
 
     /*********************************/
     /* Flash Configuration Functions */
-- 
2.27.0

