From f1255dacc2c5ebd8f8a062bca7b5a7efa642937e Mon Sep 17 00:00:00 2001
From: pennam <m.pennasilico@arduino.cc>
Date: Thu, 6 May 2021 14:29:46 +0200
Subject: [PATCH 11/87] Portenta add missing includes to resources.h

---
 .../resources/firmware/COMPONENT_4343W_FS/resources.h          | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/COMPONENT_WHD/resources/firmware/COMPONENT_4343W_FS/resources.h b/targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/COMPONENT_WHD/resources/firmware/COMPONENT_4343W_FS/resources.h
index 69d21d5b36..01bad9226f 100644
--- a/targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/COMPONENT_WHD/resources/firmware/COMPONENT_4343W_FS/resources.h
+++ b/targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/COMPONENT_WHD/resources/firmware/COMPONENT_4343W_FS/resources.h
@@ -18,6 +18,9 @@
 #ifndef INCLUDED_RESOURCES_H_
 #define INCLUDED_RESOURCES_H_
 #include "wiced_resource.h"
+#include "wiced_filesystem.h"
+
+extern wiced_filesystem_t resource_fs_handle;
 
 extern const resource_hnd_t wifi_firmware_image;
 extern const unsigned char wifi_firmware_image_data[421098];
-- 
2.30.2

