From 094ccdb7e97ed79981211a1f156e20a89fa8c21e Mon Sep 17 00:00:00 2001
From: Martino Facchin <m.facchin@arduino.cc>
Date: Tue, 7 Jan 2020 23:21:56 +0100
Subject: [PATCH 12/31] STM32H747: don't start CM4 if boot bit is not set

---
 targets/TARGET_STM/mbed_overrides.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/targets/TARGET_STM/mbed_overrides.c b/targets/TARGET_STM/mbed_overrides.c
index f8c19b6991..e8a4e680b4 100644
--- a/targets/TARGET_STM/mbed_overrides.c
+++ b/targets/TARGET_STM/mbed_overrides.c
@@ -129,11 +129,11 @@ void mbed_sdk_init()
         LL_HSEM_1StepLock(HSEM, CFG_HW_STOP_MODE_SEMID);
         /*Release HSEM in order to notify the CPU2(CM4)*/
         LL_HSEM_ReleaseLock(HSEM, CFG_HW_STOP_MODE_SEMID, 0);
+        /* wait until CPU2 wakes up from stop mode */
+        while (LL_RCC_D2CK_IsReady() == 0);
     } else {
-        LL_RCC_ForceCM4Boot();
+        //LL_RCC_ForceCM4Boot();
     }
-    /* wait until CPU2 wakes up from stop mode */
-    while (LL_RCC_D2CK_IsReady() == 0);
 #endif /* CORE_M4 */
 #else /* Single core */
     // Update the SystemCoreClock variable.
-- 
2.27.0

