From d796c56c56a7c65964b50e5d18e0bb10c18df6c0 Mon Sep 17 00:00:00 2001
From: Martino Facchin <m.facchin@arduino.cc>
Date: Mon, 8 Mar 2021 10:07:26 +0100
Subject: [PATCH 40/87] RP2040: usb: make all unhandled conditions a no-op

---
 .../TARGET_RP2040/USBPhy_RP2040.cpp              | 16 ++++------------
 1 file changed, 4 insertions(+), 12 deletions(-)

diff --git a/targets/TARGET_RASPBERRYPI/TARGET_RP2040/USBPhy_RP2040.cpp b/targets/TARGET_RASPBERRYPI/TARGET_RP2040/USBPhy_RP2040.cpp
index c88bdcc573..ff24a6dcd6 100644
--- a/targets/TARGET_RASPBERRYPI/TARGET_RP2040/USBPhy_RP2040.cpp
+++ b/targets/TARGET_RASPBERRYPI/TARGET_RP2040/USBPhy_RP2040.cpp
@@ -295,23 +295,17 @@ bool USBPhyHw::endpoint_add(usb_ep_t endpoint, uint32_t max_packet, usb_ep_type_
 
 void USBPhyHw::endpoint_remove(usb_ep_t endpoint)
 {
-    // Halt here, unhandled
-    volatile int going = true;
-    while(going);
+
 }
 
 void USBPhyHw::endpoint_stall(usb_ep_t endpoint)
 {
-    // Halt here, unhandled
-    volatile int going = true;
-    while(going);
+
 }
 
 void USBPhyHw::endpoint_unstall(usb_ep_t endpoint)
 {
-    // Halt here, unhandled
-    volatile int going = true;
-    while(going);
+
 }
 
 bool USBPhyHw::endpoint_read(usb_ep_t endpoint, uint8_t *data, uint32_t size)
@@ -385,9 +379,7 @@ bool USBPhyHw::endpoint_write(usb_ep_t endpoint, uint8_t *data, uint32_t size)
 
 void USBPhyHw::endpoint_abort(usb_ep_t endpoint)
 {
-    // Unhandled, halt here
-    volatile int going = true;
-    while(going);
+
 }
 
 void USBPhyHw::process()
-- 
2.33.1

