From 13135d4d0116b07f51d16fdfa57e2b8308d5fed2 Mon Sep 17 00:00:00 2001
From: Martino Facchin <m.facchin@arduino.cc>
Date: Mon, 15 Mar 2021 15:22:43 +0100
Subject: [PATCH 46/87] RP2040: nano connect: fix pinmap

---
 .../TARGET_NANO_RP2040_CONNECT/PinNames.h     | 25 ++++++++++++++++---
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/targets/TARGET_RASPBERRYPI/TARGET_RP2040/TARGET_NANO_RP2040_CONNECT/PinNames.h b/targets/TARGET_RASPBERRYPI/TARGET_RP2040/TARGET_NANO_RP2040_CONNECT/PinNames.h
index e84e89440c..bad8f487c6 100644
--- a/targets/TARGET_RASPBERRYPI/TARGET_RP2040/TARGET_NANO_RP2040_CONNECT/PinNames.h
+++ b/targets/TARGET_RASPBERRYPI/TARGET_RP2040/TARGET_NANO_RP2040_CONNECT/PinNames.h
@@ -44,18 +44,35 @@ typedef enum {
     ADC_TEMP = 0xF0,
     ADC_VREF = 0xF1,
 
+#ifndef ARDUINO_ARCH_MBED
+    D0          = p1,
+    D1          = p0,
+    D2          = p25,
+    D3          = p15,
+    D4          = p16,
+    D5          = p17,
+    D6          = p18,
+    D7          = p19,
+    D8          = p20,
+    D9          = p21,
+    D10         = p5,
+    D11         = p7,
+    D12         = p4,
+    D13         = p6,
+
     A0          = 26,
     A1          = 27,
     A2          = 28,
     A3          = 29,
+#endif
 
     USBTX = p0,
     USBRX = p1,
 
-    LED1 = p25,
-    LED2 = p25,
-    LED3 = p25,
-    LED4 = p25,
+    LED1 = p6,
+    LED2 = p6,
+    LED3 = p6,
+    LED4 = p6,
 
     // Not connected
     NC = (int)0xFFFFFFFF
-- 
2.33.1

