From 17a80871fb04805452e3daffa6bde3d5d116d6ac Mon Sep 17 00:00:00 2001
From: giulcioffi <g.cioffi@arduino.cc>
Date: Mon, 15 Mar 2021 12:45:15 +0100
Subject: [PATCH 02/16] RP2040: Fix write/read command detection

---
 .../TARGET_RASPBERRYPI/TARGET_RP2040/i2c_api.c    | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/targets/TARGET_RASPBERRYPI/TARGET_RP2040/i2c_api.c b/targets/TARGET_RASPBERRYPI/TARGET_RP2040/i2c_api.c
index 219fafb0fb..f176b50728 100644
--- a/targets/TARGET_RASPBERRYPI/TARGET_RP2040/i2c_api.c
+++ b/targets/TARGET_RASPBERRYPI/TARGET_RP2040/i2c_api.c
@@ -168,17 +168,20 @@ void i2c_slave_mode(i2c_t *obj, int enable_slave)
  */
 int i2c_slave_receive(i2c_t *obj)
 {
-    //DEBUG_PRINTF("i2c_slave_receive\r\n");
     int retValue = NoData;
 
-    int command = (obj->dev->hw->data_cmd & 0x00000100) >> 8;
+    int rd_req = (obj->dev->hw->raw_intr_stat & I2C_IC_RAW_INTR_STAT_RD_REQ_BITS) >> 5;
 
-    if (command == I2C_IC_DATA_CMD_CMD_VALUE_WRITE) {
-        retValue = WriteAddressed;
+    if (rd_req == I2C_IC_RAW_INTR_STAT_RD_REQ_VALUE_ACTIVE) {
+        DEBUG_PRINTF("Read addressed\r\n");
+        return ReadAddressed;
     }
 
-    if (command == I2C_IC_DATA_CMD_CMD_VALUE_READ) {
-        retValue = ReadAddressed;
+    int wr_req = (obj->dev->hw->status & I2C_IC_STATUS_RFNE_BITS) >> 3;
+
+    if (wr_req == I2C_IC_STATUS_RFNE_VALUE_NOT_EMPTY) {
+        DEBUG_PRINTF("Write addressed\r\n");
+        return WriteAddressed;
     }
 
     return (retValue);
-- 
2.30.2

