From bbdc9efdc824d23a3e573ef11c0f50117359e030 Mon Sep 17 00:00:00 2001
From: Martino Facchin <m.facchin@arduino.cc>
Date: Tue, 10 Aug 2021 12:09:30 +0200
Subject: [PATCH 70/87] RP2040: enable led indication on mbed_die()

---
 targets/TARGET_RASPBERRYPI/TARGET_RP2040/objects.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/targets/TARGET_RASPBERRYPI/TARGET_RP2040/objects.h b/targets/TARGET_RASPBERRYPI/TARGET_RP2040/objects.h
index 280f166d4d..f5f44a58f4 100644
--- a/targets/TARGET_RASPBERRYPI/TARGET_RP2040/objects.h
+++ b/targets/TARGET_RASPBERRYPI/TARGET_RP2040/objects.h
@@ -129,4 +129,6 @@ typedef struct serial_s serial_t;
 }
 #endif
 
+#define LED1   LED1
+
 #endif
-- 
2.33.1

