From 9179831ca7fe22dcf3c095fd07232e495d2beaa9 Mon Sep 17 00:00:00 2001
From: giulcioffi <g.cioffi@arduino.cc>
Date: Thu, 3 Feb 2022 14:35:55 +0100
Subject: [PATCH 112/204] PwmOut: set pwm period before duty cycle

---
 drivers/source/PwmOut.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/source/PwmOut.cpp b/drivers/source/PwmOut.cpp
index cc717f8e99..53e6f84f10 100644
--- a/drivers/source/PwmOut.cpp
+++ b/drivers/source/PwmOut.cpp
@@ -142,8 +142,8 @@ void PwmOut::resume()
     core_util_critical_section_enter();
     if (!_initialized) {
         PwmOut::init();
-        PwmOut::write(_duty_cycle);
         PwmOut::period_us(_period_us);
+        PwmOut::write(_duty_cycle);
     }
     core_util_critical_section_exit();
 }
-- 
2.39.1

