From b3b0bdb971db0225a24bf9388f9a85c78c960782 Mon Sep 17 00:00:00 2001
From: pennam <m.pennasilico@arduino.cc>
Date: Wed, 2 Mar 2022 11:31:01 +0100
Subject: [PATCH 138/204] Initialize sss_object and sss_keystore to nullptr

---
 connectivity/netsocket/include/netsocket/TLSSocketWrapper.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/connectivity/netsocket/include/netsocket/TLSSocketWrapper.h b/connectivity/netsocket/include/netsocket/TLSSocketWrapper.h
index 13365428fd..2dc3b4b000 100644
--- a/connectivity/netsocket/include/netsocket/TLSSocketWrapper.h
+++ b/connectivity/netsocket/include/netsocket/TLSSocketWrapper.h
@@ -366,8 +366,8 @@ private:
 #endif
 
 #if defined(COMPONENT_SE050) && defined(MBEDTLS_ECDH_ALT) && SSS_HAVE_ALT_SSS
-    sss_object_t *_sss_key_pair_ptr;
-    sss_key_store_t *_sss_ks_ptr;
+    sss_object_t *_sss_key_pair_ptr = nullptr;
+    sss_key_store_t *_sss_ks_ptr = nullptr;
 #endif
 
     DRBG_CTX _drbg;
-- 
2.39.1

