From 4b805b51458f021a5e8b2f8cbe1f8f783b34cb1e Mon Sep 17 00:00:00 2001
From: Martino Facchin <m.facchin@arduino.cc>
Date: Tue, 10 May 2022 18:07:10 +0200
Subject: [PATCH 155/204] ST: CAN: allow sending CAN remote frames

---
 targets/TARGET_STM/can_api.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/targets/TARGET_STM/can_api.c b/targets/TARGET_STM/can_api.c
index b18b9cb747..c579c19a7e 100644
--- a/targets/TARGET_STM/can_api.c
+++ b/targets/TARGET_STM/can_api.c
@@ -408,7 +408,11 @@ int can_write(can_t *obj, CAN_Message msg, int cc)
         TxHeader.IdType = FDCAN_EXTENDED_ID;
     }
 
-    TxHeader.TxFrameType = FDCAN_DATA_FRAME;
+    if (msg.type == CANData) {
+        TxHeader.TxFrameType = FDCAN_DATA_FRAME;
+    } else {
+        TxHeader.TxFrameType = FDCAN_REMOTE_FRAME;
+    }
     TxHeader.DataLength = msg.len << 16;
     TxHeader.ErrorStateIndicator = FDCAN_ESI_ACTIVE;
     TxHeader.BitRateSwitch = FDCAN_BRS_OFF;
-- 
2.39.1

