From 77fb09b520e9914d457882ef63e4a795d3e87529 Mon Sep 17 00:00:00 2001
From: Martino Facchin <m.facchin@arduino.cc>
Date: Fri, 23 Sep 2022 09:51:20 +0200
Subject: [PATCH 177/204] QSPI: make _qspi object protected to allow
 subclassing

---
 .../COMPONENT_QSPIF/include/QSPIF/QSPIFBlockDevice.h            | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/storage/blockdevice/COMPONENT_QSPIF/include/QSPIF/QSPIFBlockDevice.h b/storage/blockdevice/COMPONENT_QSPIF/include/QSPIF/QSPIFBlockDevice.h
index 2903ecad32..695a396197 100644
--- a/storage/blockdevice/COMPONENT_QSPIF/include/QSPIF/QSPIFBlockDevice.h
+++ b/storage/blockdevice/COMPONENT_QSPIF/include/QSPIF/QSPIFBlockDevice.h
@@ -340,9 +340,11 @@ private:
         QSPIF_BP_CLEAR_SR, // Clear protection bits in status register 1
     };
 
+protected:
     // QSPI Driver Object
     mbed::QSPI _qspi;
 
+private:
     // Static List of different QSPI based Flash devices csel that already exist
     // Each QSPI Flash device csel can have only 1 QSPIFBlockDevice instance
     // _devices_mutex is used to lock csel list - only one QSPIFBlockDevice instance per csel is allowed
-- 
2.39.1

